### PR TITLE
Python Backend Test Submission

### PR DESCRIPTION
**Pull request by dhruvsingh**
Tasks
- [TASK 1](https://github.com/dhruvsingh/backend-python-test/commit/91edc9901b5b5e7c747c483db1748755ac98520b)
    - Add frontend required validation
    - Add backend validation for description field
- [TASK 2](https://github.com/dhruvsingh/backend-python-test/commit/43735ea123a2c35ac61cd81e3731e776a6829bee)
    - Add field to todo model (completed: bool)
    - Add migration script to resources/(alters current table to add completed field)
    - Update main.py to be able to run ```python main.py migrate``` from terminal to migrate data

- [TASK 3](https://github.com/dhruvsingh/backend-python-test/commit/c6fc30dfe9a02c9b7b6d604ed248a5fc31cec718)
    - Add API endpoint to respond with json data for todo detail
    - Users can see anyone's todo in this commit; has been handled under additional tasks

- [TASK 4](https://github.com/dhruvsingh/backend-python-test/commit/566ca04c295f6404f8450dea414392b7a5efec37)
    - Use flash, flask inbuilt package for showing confirmation messages for adding, not found, deleting and completing a todo.

- [TASK 5](https://github.com/dhruvsingh/backend-python-test/commit/f718ea378bf38070c313bc1e6f2a3b451ecc014c)
    - Quick approach - Add datatables to paginate todos in the frontend

- [TASK 6](https://github.com/dhruvsingh/backend-python-test/commit/95e6dea7881141c02452654367a354e27c878e51)
    - Add flask sqlalchemy to implement database ORM
    - Add models for user and todo
    - Fix html to update user_id reference to user

[Additional Tasks](https://github.com/dhruvsingh/backend-python-test/commit/2c80b0d53140f99de131dd84f2bd792415cc67f3)
- Security (fixed as per app [requirements](https://github.com/AlayaCare/backend-python-test#application))
    - Users should see their own todo detail (both json and frontend)
    - Users should be able to delete their own todo
    - Users should be able to see their todo in json

- [Html](https://github.com/dhruvsingh/backend-python-test/commit/726b0613d4098dea7d2dfc76d3d82e17111001eb)
    - Use bootstrap form for adding description
    - Add title for hovering on icons
    - Disable the ability to mark todo as completed if it is already completed

## Enhancements possible further
    - Using flask-fixtures to import data (partially done; users data being imported already, some issue with importing data for todos)
    - Backend pagination and then frontend js/jquery to make paginated calls to the backend.
    - Extract out db.session.add() and db.session.commit()
    - pep8ify views.py
    - chmod +x main.py (to write one less word :) )

### CODE REVIEW

File: alayatodo/views.py [hunk 1]
```diff
@@ -65,18 +80,90 @@ def todos():
...
+        'message': 'OK',
+        'status_code': 200,
+        'results': []
+    }
+
+    if not todo:
+        data['status_code'] = 404
+        data['message'] = 'Not Found'
+    else:
+        data['results'].append({key: todo[key] for key in todo.keys()})
```
Comment (m-a-choquette) [inline alayatodo/views.py]: There is an id in the route, the json response should not be a collection of todos, but a single todo object instead.

todo is an instance of the Todo model and calling `.keys()` should raise raise an `AttributeError`.

File: alayatodo/views.py [hunk 2]
```diff
@@ -65,18 +80,90 @@ def todos():
...
+        db.session.commit()
+
+        flash("Todo deleted succesfully.")
+    else:
+        flash("Todo not found.")
+
     return redirect('/todo')
+
+
+@app.route('/todo/complete/<id>', methods=['POST'])
```
Comment (m-a-choquette) [inline alayatodo/views.py]: The type of the `id` parameter can be specified in the route:  `<int:id>`

File: alayatodo/views.py [hunk 3]
```diff
@@ -65,18 +80,90 @@ def todos():
...
+def todo_complete(id):
+    if not session.get('logged_in'):
+        return redirect('/login')
```
Comment (m-a-choquette) [inline alayatodo/views.py]: This authentication check could be moved into a decorator

**Comment by m-a-choquette** [comment]
Hi, 
thanks for submitting your test,
have a nice day!

**Comment by dhruvsingh** [comment]
Hi Marc,

Thank you for taking out the time to review my changes and sharing the areas that it can be improved on. Really appreciate it.

I'll work on them and do the needful.

**Pull request closed without merging**